Problems with apoc.diff.nodes and Neo4j 4.2

## Guidelines

Please note that GitHub issues are only meant for bug reports/feature requests. If you have questions on how to use APOC, please ask on [the Neo4j Discussion Forum](https://community.neo4j.com/c/neo4j-graph-platform/procedures-apoc) instead of creating an issue here.

After updating Neo4j 3.5.16 to 4.2 i have problems with the apoc function apoc.diff.nodes.
Mr. Armbruster advised me to open a ticket for this issue.
Due to the changed transactional behavior apoc.diff.nodes can not work with virtual nodes. 


## Expected Behavior (Mandatory)


## Actual Behavior (Mandatory)
Here the error message:
org.neo4j.driver.v1.exceptions.ClientException: Failed to invoke function `apoc.diff.nodes`: Caused by: org.neo4j.internal.kernel.api.exceptions.EntityNotFoundException: Unable to load NODE with id -4. 
at org.neo4j.driver.internal.util.Futures.blockingGet(Futures.java:123) ~[neo4j-java-driver-1.7.5.jar:1.7.5-d56154a7e96ff71615b894ee9a15d08c370691e6] 
at org.neo4j.driver.internal.ExplicitTransaction.close(ExplicitTransaction.java:117) ~[neo4j-java-driver-1.7.5.jar:1.7.5-d56154a7e96ff71615b894ee9a15d08c370691e6] 
at de.guh.plugin.neo4j.Graph.execute(Graph.java:203) [inchorus-neo4j-plugin-2.0.12.jar:?] at de.guh.gadget.NodeCollection.sendCypherStatement(NodeCollection.java:982) [classes/:?] 
at de.guh.gadget.NodeCollection.runNodeCollection(NodeCollection.java:407) [classes/:?] at de.guh.gadget.Collector.runCollector(Collector.java:1096) [classes/:?] 
at de.guh.gadget.Collector.execute(Collector.java:2048) [classes/:?] at it.sauronsoftware.cron4j.TaskExecutor$Runner.run(Unknown Source) [cron4j-2.2.5.jar:?] 
at java.lang.Thread.run(Thread.java:834) [?:?]


## How to Reproduce the Problem

### Simple Dataset (where it's possibile)

```
UNWIND $data as object OPTIONAL MATCH (preexists:User:SecurityPrincipal{uniqueid:object.uniqueid, collectionid: object.collectionid}) MERGE (source:User:SecurityPrincipal{uniqueid:object.uniqueid, collectionid: object.collectionid}) WITH object, source, preexists, apoc.diff.nodes(source,apoc.create.vNode(['User:SecurityPrincipal'], object)) as delta OPTIONAL MATCH (source)<-[targetrelation_1_1:OWNS_USER]-(:ActiveDirectoryDomain {systemid: '1'}) CALL apoc.do.when( targetrelation_1_1 IS NULL,  " OPTIONAL MATCH (source)<-[targetrelation_1_2:OWNS_USER]-(:ActiveDirectoryDomain)  DELETE targetrelation_1_2  WITH source, object  OPTIONAL MATCH (targetnode_1_1:ActiveDirectoryDomain {systemid: '1'})  CALL apoc.do.when(  targetnode_1_1 IS NOT NULL , 'MERGE (source1)<-[:OWNS_USER]-(targetnode_1_1) RETURN 1' , 'RETURN 1' , {source1:source, targetnode_1_1:targetnode_1_1}  )  YIELD value RETURN 1  " ,  'RETURN 1' , {source:source, object:object}  )  YIELD value AS ignored1 OPTIONAL MATCH (source)<-[targetrelation_2_1:HAS_SECURITYPRINCIPAL]-(:Person {username:object.name}) CALL apoc.do.when( targetrelation_2_1 IS NULL ,  " OPTIONAL MATCH (source)<-[targetrelation_2_2:HAS_SECURITYPRINCIPAL]-(:Person)  DELETE targetrelation_2_2  WITH source, object  OPTIONAL MATCH (targetnode_2_1:Person {username:object.name})  CALL apoc.do.when(  targetnode_2_1 IS NOT NULL , 'MERGE (source1)<-[:HAS_SECURITYPRINCIPAL]-(targetnode_2_1) RETURN 1' , 'RETURN 1' , {source1:source, targetnode_2_1:targetnode_2_1}  )  YIELD value RETURN 1 " , 'RETURN 1' , {source:source, object:object}  )  YIELD value AS ignored2 SET source += object RETURN distinct source.uniqueid AS uniqueid, delta, preexists IS NULL AS hasBeenCreated

```


### Steps (Mandatory)

  1.
  1.
  1.

## Screenshots (where it's possibile)

## Specifications (Mandatory)

Currently used versions

### Versions

  - OS:
  - Neo4j: 4.2
  - Neo4j-Apoc: 4.2.0.1

--- a/core/src/main/java/apoc/util/Util.java
+++ b/core/src/main/java/apoc/util/Util.java
@@ -3,6 +3,8 @@ package apoc.util;
 import apoc.Pools;
 import apoc.convert.Convert;
 import apoc.export.util.CountingInputStream;
+import apoc.result.VirtualNode;
+import apoc.result.VirtualRelationship;
 import org.apache.commons.io.IOUtils;
 import org.apache.commons.lang.StringUtils;
 import org.eclipse.collections.api.iterator.LongIterator;
@@ -849,13 +851,12 @@ public class Util {
                 .collect(Collectors.toMap(e -> e.getKey(), e -> e.getValue()));
     }
 
-
     public static Node rebind(Transaction tx, Node node) {
-        return tx.getNodeById(node.getId());
+         return node instanceof VirtualNode ? node : tx.getNodeById(node.getId());
     }
 
     public static Relationship rebind(Transaction tx, Relationship rel) {
-        return tx.getRelationshipById(rel.getId());
+         return rel instanceof VirtualRelationship ? rel : tx.getRelationshipById(rel.getId());
     }
 
     public static Entity rebind(Transaction tx, Entity e) {

--- a/core/src/test/java/apoc/diff/DiffTest.java
+++ b/core/src/test/java/apoc/diff/DiffTest.java
@@ -1,5 +1,6 @@
 package apoc.diff;
 
+import apoc.create.Create;
 import apoc.util.TestUtil;
 import org.junit.BeforeClass;
 import org.junit.ClassRule;
@@ -31,7 +32,7 @@ public class DiffTest {
 
     @BeforeClass
     public static void setup() throws Exception {
-        TestUtil.registerProcedure(db, Diff.class);
+        TestUtil.registerProcedure(db, Diff.class, Create.class);
 
         try (Transaction tx = db.beginTx()) {
             node1 = tx.createNode(Label.label("Node1"));
@@ -53,10 +54,6 @@ public class DiffTest {
 
     @Test
     public void nodesSame() {
-        Map<String, Object> params = new HashMap<>();
-        params.put("leftNode", node1);
-        params.put("rightNode", node1);
-
         Map<String, Object> result =
                 db.executeTransactionally(
                         "MATCH (node:Node1) RETURN apoc.diff.nodes(node, node) as diff", new HashMap<>(),
@@ -80,10 +77,6 @@ public class DiffTest {
 
     @Test
     public void nodesDiffering() {
-        Map<String, Object> params = new HashMap<>();
-        params.put("leftNode", node2);
-        params.put("rightNode", node3);
-
         Map<String, Object> result =
                 db.executeTransactionally(
                         "MATCH (leftNode:Node2), (rightNode:Node3) RETURN apoc.diff.nodes(leftNode, rightNode) as diff", new HashMap<>(),
@@ -107,6 +100,59 @@ public class DiffTest {
         HashMap<String, Object> inCommon = (HashMap<String, Object>) result.get("inCommon");
         assertEquals(1, inCommon.size());
         assertEquals("val1", inCommon.get("prop1"));
+    }
+
+    @Test
+    public void shouldBeDiffWithVirtualNodes() {
+        String query = "WITH apoc.create.vNode(['Node2'], {prop1: 'val1', prop2: 2, prop4: 'four'}) AS nodeA, " +
+                "apoc.create.vNode(['Node3'], {prop1: 'val1', prop3: '3', prop4: 'for'}) AS nodeB " +
+                "RETURN apoc.diff.nodes(nodeA, nodeB) as diff";
+        Map<String, Object> result =
+                db.executeTransactionally(query, Map.of(),
+                        r -> Iterators.single(r.columnAs("diff")));
+        assertNotNull(result);
+
+        HashMap<String, Object> leftOnly = (HashMap<String, Object>) result.get("leftOnly");
+        assertEquals(1, leftOnly.size());
+        assertEquals(2L, leftOnly.get("prop2"));
+
+        HashMap<String, Object> rightOnly = (HashMap<String, Object>) result.get("rightOnly");
+        assertEquals(1, rightOnly.size());
+        assertEquals("3", rightOnly.get("prop3"));
+
+        HashMap<String, HashMap<String, Object>> different = (HashMap<String, HashMap<String, Object>>) result.get("different");
+        assertEquals(1, different.size());
+        HashMap<String, Object> pairs = different.get("prop4");
+        assertEquals("four", pairs.get("left"));
+        assertEquals("for", pairs.get("right"));
+
+        HashMap<String, Object> inCommon = (HashMap<String, Object>) result.get("inCommon");
+        assertEquals(1, inCommon.size());
+        assertEquals("val1", inCommon.get("prop1"));
+    }
+
+    @Test
+    public void shouldBeSameWithVirtualNodes() {
+        String query = "WITH apoc.create.vNode(['Node1'], {prop1: 'val1', prop2: 2}) AS node " +
+                "RETURN apoc.diff.nodes(node, node) as diff";
+        Map<String, Object> result =
+                db.executeTransactionally(query, Map.of(),
+                        r -> Iterators.single(r.columnAs("diff")));
+        assertNotNull(result);
+
+        HashMap<String, Object> leftOnly = (HashMap<String, Object>) result.get("leftOnly");
+        assertTrue(leftOnly.isEmpty());
+
+        HashMap<String, Object> rightOnly = (HashMap<String, Object>) result.get("rightOnly");
+        assertTrue(rightOnly.isEmpty());
+
+        HashMap<String, Object> different = (HashMap<String, Object>) result.get("different");
+        assertTrue(different.isEmpty());
+
+        HashMap<String, Object> inCommon = (HashMap<String, Object>) result.get("inCommon");
+        assertEquals(2, inCommon.size());
+        assertEquals("val1", inCommon.get("prop1"));
+        assertEquals(2L, inCommon.get("prop2"));
     }
 
 }

